### PR TITLE
Add index_of plugin

### PR DESCRIPTION
**Pull request by cidrblock**

### CODE REVIEW

File: docs/ansible.utils.index_of_lookup.rst [hunk 1]
```diff
@@ -0,0 +1,503 @@
...
+    - name: Find the index of all firewalls, use in a loop, as a filter
+      debug:
+        msg: "The type of {{ device_type }} at index {{ item }} has name {{ data[item].name }}."
+      loop: "{{ data|ansible.netcommon.index_of('eq', device_type, 'type') }}"
+      vars:
+        device_type: firewall
+
+    # TASK [Find the index of all firewalls, use in a loop] **********************
+    # ok: [sw01] => (item=2) =>
+    #   msg: The type of firewall at index 2 has name fw01.
```
Comment (ganeshrn) [inline docs/ansible.utils.index_of_lookup.rst:316]: Based on example replace `fw01` to `fw01.example.corp`

File: docs/ansible.utils.index_of_lookup.rst [hunk 2]
```diff
@@ -0,0 +1,503 @@
...
+    # ok: [sw01] => (item=3) =>
+    #   msg: The type of firewall at index 3 has name fw02.
```
Comment (ganeshrn) [inline docs/ansible.utils.index_of_lookup.rst:318]: similar to above

File: plugins/lookup/index_of.py [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+"""
+from __future__ import absolute_import, division, print_function
+
+__metaclass__ = type
+
+
+DOCUMENTATION = """
+    lookup: index_of
+    author: Bradley Thornton (@cidrblock)
+    version_added: "1.4"
```
Comment (ganeshrn) [inline plugins/lookup/index_of.py:18]: this should be `1.0.0`

File: plugins/lookup/index_of.py [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+    short_description: Find the indicies of items in a list matching some criteria
+    description:
+        - This lookup returns a list of indicies of items matching some criteria in a list.
```
Comment (ganeshrn) [inline plugins/lookup/index_of.py:21]: If I understand correctly the return is of type list if `wantlist` is set to `True`. Should the description change accordingly.

**Review by ganeshrn — APPROVED** [review]

**Review by ganeshrn — COMMENTED** [review]
Apart from minor doc changes, PR looks good to me.

**Comment by cidrblock** [comment]
Leaving a note here for future me, it seems the `gate` label may not work when a reviewer was a team: https://github.com/cli/cli/issues/1314

**Comment by pabelanger** [comment]
@cidrblock we can onboard to zuul today

**Pull request merged**